src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/render/ModuleBreadcrumbs.kt

```diff
@@ -18,91 +18,178 @@
-    private var lastPosY = 0.0
-    private var lastPosZ = 0.0
-
-    override fun enable() {
-        synchronized(positions) {
-            positions.addAll(listOf(player.x, player.eyeY, player.z))
-            positions.addAll(listOf(player.x, player.y, player.z))
-        }
-    }
+    private val positions = mutableMapOf<Entity, MutableList<TrailPart>>()
+    private val lastPositions = mutableMapOf<Entity, Array<Double>>()
 
     override fun disable() {
-        synchronized(positions) {
-            positions.clear()
-        }
+        clear()
     }
 
     val renderHandler = handler<WorldRenderEvent> { event ->
+        val time = System.currentTimeMillis()
+        update(time)
+
         val matrixStack = event.matrixStack
         val color = if (colorRainbow) rainbow() else color
 
-        synchronized(positions) {
-            renderEnvironmentForWorld(matrixStack) {
-                withColor(color) {
-                    val lines = makeLines(positions, event.partialTicks)
-                        .map { relativeToCamera(it).toVec3() }
-                        .toTypedArray()
-                    @Suppress("SpreadOperator")
-                    drawLineStrip(*lines)
-                }
+        renderEnvironmentForWorld(matrixStack) {
+            if (height > 0) {
+                RenderSystem.disableCull()
+            }
+
+            positions.values.forEach { doubles ->
+                val lines = makeAndVerifyVertexList(time, doubles)
+                    .map { Pair(relativeToCamera(it.first).toVec3(), it.second) }
+                    .toTypedArray()
+
+                @Suppress("SpreadOperator")
+                draw(matrixStack, color, *lines)
+            }
+
+            if (height > 0) {
+                RenderSystem.enableCull()
             }
         }
     }
 
-    private fun makeLines(positions: List<Double>, tickDelta: Float): List<Vec3d> {
-        val mutableList = mutableListOf<Vec3d>()
-        for (i in 0 until positions.size / 3 - 1) {
-            mutableList += Vec3d(positions[i * 3], positions[i * 3 + 1], positions[i * 3 + 2])
+    private fun makeAndVerifyVertexList(time: Long, positions: MutableList<TrailPart>): List<Pair<Vec3d, Float>> {
+        val mutableList = mutableListOf<Pair<Vec3d, Float>>()
+        val toRemove = mutableListOf<TrailPart>()
+        val aliveDuration = alive.toLong()
+        val alpha = color.a / 255f
+
+        positions.forEach { position ->
+            val deltaTime = time - position.creationTime
+            if (deltaTime > aliveDuration) {
+                toRemove += position
+            } else {
+                val calcAlpha = if (fade) (1f - (deltaTime / aliveDuration).toFloat()) * alpha else alpha
+                mutableList += Pair(Vec3d(position.x, position.y, position.z), calcAlpha)
+            }
         }
-        mutableList += player.interpolateCurrentPosition(tickDelta)
+
+        positions.removeAll(toRemove)
         return mutableList
     }
```

Comment:
This implementation currently has no real performance issues, but it is still sloppy. You can do this in a much more elegant way, let me show you:

The problem is that currently you create a list with all the TrailParts you want to remove and for every item in `positions` the code will have to call `TrailPart.equals()` for every item in `toRemove`. It is `O(n*m)` which is theoretically very bad, but at the moment `n*m` is `< 200` so it is an okay-ish solution.

The structure you want to use is an `ArrayDeque`, it is a great fit for many situations, it can basically add/remove the first or last element really quick (`ArrayList`s can only remove the last element without reorganizing the entire list).

You can simplify this code to:
```Kotlin
// positions: ArrayDeque<TrailPart>

val timeThreshold = time - aliveDuration
var head = positions.peekFirst()

// Remove outdated positions, the positions are ordered by time (ascending)
while (head != null && head.creationTime < timeThreshold) {
    positions.removeFirst()

    head = positions.peekFirst()
}

val pointsWithAlpha = positions.map {
    val calcAlpha = if (fade) (1f - (deltaTime / aliveDuration).toFloat()) * alpha else alpha

    Pair(Vec3d(position.x, position.y, position.z), calcAlpha)
}

return pointsWithAlpha
```
This code runs 0 unnecessary operations